src/combinatorics/additive/salem_spencer.lean

```diff
@@ -0,0 +1,431 @@
+end
+
+end monoid
+
+section cancel_comm_monoid
+variables [cancel_comm_monoid α] (s : finset α) (a : α)
+
+@[simp, to_additive] lemma mul_roth_number_map_mul_left :
+  mul_roth_number (s.map $ mul_left_embedding a) = mul_roth_number s :=
+begin
+  refine le_antisymm _ _,
+  { obtain ⟨u, hus, hcard, hu⟩ := mul_roth_number_spec (s.map $ mul_left_embedding a),
+    rw subset_map_iff at hus,
+    obtain ⟨u, hus, rfl⟩ := hus,
+    rw coe_map at hu,
+    rw [←hcard, card_map],
+    exact (mul_salem_spencer_mul_left_iff.1 hu).le_mul_roth_number hus },
+  { obtain ⟨u, hus, hcard, hu⟩ := mul_roth_number_spec s,
+    have h : mul_salem_spencer (u.map $ mul_left_embedding a : set α),
+    { rw coe_map,
+      exact hu.mul_left },
+    convert h.le_mul_roth_number (map_subset_map.2 hus),
+    rw [card_map, hcard] }
+end
+
+@[simp, to_additive] lemma mul_roth_number_map_mul_right :
+  mul_roth_number (s.map $ mul_right_embedding a) = mul_roth_number s :=
+begin
+  convert mul_roth_number_map_mul_left s a,
+  ext,
+  exact mul_comm _ _,
+end
+
+end cancel_comm_monoid
+end roth_number
+
+section roth_number_nat
+variables {s : finset ℕ} {k n : ℕ}
+
+/-- The Roth number of a natural `N` is the largest integer `m` for which there is a subset of
+`range N` of size `m` with no arithmetic progression of length 3.
+Trivially, `roth_number N ≤ N`, but Roth's theorem (proved in 1953) shows that
+`roth_number N = o(N)` and the construction by Behrend gives a lower bound of the form
+`N * exp(-C sqrt(log(N))) ≤ roth_number N`.
+A significant refinement of Roth's theorem by Bloom and Sisask announced in 2020 gives
+`roth_number N = O(N / (log N)^(1+c))` for an absolute constant `c`. -/
+def roth_number_nat : ℕ →ₘ ℕ :=
+⟨λ n, add_roth_number (range n), add_roth_number.mono.comp range_mono⟩
+
+lemma roth_number_nat_def (n : ℕ) : roth_number_nat n = add_roth_number (range n) := rfl
+
+lemma roth_number_nat_le (N : ℕ) : roth_number_nat N ≤ N :=
+(add_roth_number_le _).trans (card_range _).le
+
+/-- A verbose specialization of `add_salem_spencer.le_add_roth_number`, sometimes convenient in
+practice. -/
+lemma add_salem_spencer.le_roth_number_nat (s : finset ℕ) (hs : add_salem_spencer (s : set ℕ))
+  (hsn : ∀ x ∈ s, x < n) (hsk : s.card = k) :
+  k ≤ roth_number_nat n :=
+begin
+  rw ←hsk,
+  exact hs.le_add_roth_number (λ x hx, mem_range.2 $ hsn x hx),
+end
+
+/-- The Roth number is a subadditive function. Note that by Fekete's lemma this shows that
+the limit `roth_number N / N` exists, but Roth's theorem gives the stronger result that this
+limit is actually `0`. -/
+lemma roth_number_nat_add_le (M N : ℕ) :
+  roth_number_nat (M + N) ≤ roth_number_nat M + roth_number_nat N :=
+begin
+  simp_rw roth_number_nat_def,
+  rw [range_add_eq_union, ←add_roth_number_map_add_left (range N) M],
+  exact add_roth_number_union_le _ _,
+end
+
+open asymptotics filter
+
+lemma trivial_roth_bound' : is_O_with 1 (λ N, (roth_number_nat N : ℝ)) (λ N, (N : ℝ)) at_top :=
```

Comment:
I wrote this for functions N -> R with coercions, but maybe it's better to have them as functions R -> R with a floor in the left part? The usual statement on paper avoids floor, but perhaps it's more general?

Reply:
`roth_number_nat` really is a function `ℕ → ℝ` and I can't think of a context where we would need this result for `ℝ → ℝ`. How hard is it to translate between both?

Reply:
In the more complex bounds on `roth_number_nat` I think it's not trivial to convert from the `N -> R` version to the `R -> R` version, but the converse is trivial by precomposing with the coercion

Reply:
Let's keep that for when we need them.